pkg/sloop/storemanager/storemanager.go

```diff
@@ -188,26 +193,38 @@ func doCleanup(tables typed.Tables, timeLimit time.Duration, sizeLimitBytes int,
 		metricAgeOfMaximumPartition.Set(maxPartitionAge)
 	}
 
-	var totalNumOfDeletedKeys float64 = 0
-	var totalNumOfKeysToDelete float64 = 0
+	var totalNumOfDeletedKeys int64 = 0
+	var totalNumOfKeysToDelete int64 = 0
 	anyCleanupPerformed := false
-	if cleanUpTimeCondition(minPartition, maxPartition, timeLimit) || cleanUpFileSizeCondition(stats, sizeLimitBytes) {
-		partStart, partEnd, err := untyped.GetTimeRangeForPartition(minPartition)
-		glog.Infof("GC removing partition %q with data from %v to %v (err %v)", minPartition, partStart, partEnd, err)
-		var errMessages []string
-		for _, tableName := range tables.GetTableNames() {
-			prefix := fmt.Sprintf("/%s/%s", tableName, minPartition)
-			start := time.Now()
-			err, numOfDeletedKeys, numOfKeysToDelete := untyped.DeleteKeysWithPrefix([]byte(prefix), tables.Db(), deletionBatchSize)
-			metricGcDeletedNumberOfKeysByTable.WithLabelValues(fmt.Sprintf("%v", tableName)).Set(numOfDeletedKeys)
-			totalNumOfDeletedKeys += numOfDeletedKeys
-			totalNumOfKeysToDelete += numOfKeysToDelete
-			elapsed := time.Since(start)
-			glog.Infof("Call to DeleteKeysWithPrefix(%v) took %v and removed %f keys with error: %v", prefix, elapsed, numOfDeletedKeys, err)
-			if err != nil {
-				errMessages = append(errMessages, fmt.Sprintf("failed to cleanup with min key: %s, elapsed: %v,err: %v,", prefix, elapsed, err))
-			}
-			anyCleanupPerformed = true
+	minPartitionStartTime, err := untyped.GetTimeForPartition(minPartition)
+	if err != nil {
+		return false, 0, 0, err
+	}
+
+	var numOfKeysToDeleteForFileSizeCondition int64 = 0
+	isFileSizeConditionMet, garbageCollectionRatio := cleanUpFileSizeCondition(stats, sizeLimitBytes, gcThreshold)
+
+	if isFileSizeConditionMet {
+		numOfKeysToDeleteForFileSizeCondition = getNumberOfKeysToDelete(tables.Db(), garbageCollectionRatio)
+	}
+
+	beforeGCTime := time.Now()
+	for cleanUpTimeCondition(minPartition, maxPartition, timeLimit) || numOfKeysToDeleteForFileSizeCondition > 0 {
+
+		numOfDeletedKeysforPrefix, numOfKeysToDeleteForPrefix, errMessages := deletePartition(minPartition, tables, deletionBatchSize)
+		totalNumOfDeletedKeys += int64(numOfDeletedKeysforPrefix)
+		totalNumOfKeysToDelete += int64(numOfKeysToDeleteForPrefix)
+		anyCleanupPerformed = true
+		minPartitionStartTime = minPartitionStartTime.Add(untyped.GetPartitionDuration())
+		minPartition = untyped.GetPartitionId(minPartitionStartTime)
+
+		minPartitionAge, err := untyped.GetAgeOfPartitionInHours(minPartition)
+		if err == nil {
```

Comment:
nit - I'm thinking team style would rearrange the error condition first, perhaps like:
`if err != nil || minPartitionAge < 0 { return false, ... }`
`metricAgeOfMinimumPartition.Set()`